File: cheque_manager/cheque_manager/report/cheques/cheques.js
// Copyright (c) 2024, Mohammed Nasser and contributors
// For license information, please see license.txt

frappe.query_reports["Cheques"] = {
	filters: [

		{
			label: __("Company"),
			fieldname: "company",
			fieldtype: "Link",
			options: "Company",
			default: frappe.defaults.get_user_default("Company"),
		},
		{
			fieldname: "from_date",
			label: __("From Date"),
			fieldtype: "Date",
			reqd: 1,
			default: frappe.defaults.get_default("year_start_date"),
		},
		{
			fieldname: "to_date",
			label: __("To Date"),
			fieldtype: "Date",
			reqd: 1,
			default: frappe.defaults.get_default("year_end_date"),
		},
		{
			fieldname: "cheque_type",
			label: __("Cheque Type"),
			fieldtype: "MultiSelectList",
			options: "Mode of Payment",
			get_data: function (txt) {
				return frappe.db.get_link_options("Mode of Payment", txt, {
					type: "Bank",
					custom_is_cheque: 1
				});
			},
		},
		{
			fieldname: "bank_name",
			label: __("Bank Name"),
			fieldtype: "MultiSelectList",
			options: "Account",
			get_data: function (txt) {
				return frappe.db.get_link_options("Account", txt, {
					account_type: "Bank",
					is_group: 0
				});
			},
		},
		{
			fieldname: "beneficiary_name",
			label: __("Beneficiary Name"),
			fieldtype: "Data",
		},
		{
			fieldname: "status",
			label: __("Status"),
			fieldtype: "Select",
			options: [
				"",
				{
					"label": __("Draft"),
					"value": "Draft"
				},
				{
					"label": __("Submitted"),
					"value": "Submitted"
				},
				{
					"label": __("Cancelled"),
					"value": "Cancelled"
				},
				{
					"label": __("In Cheque Treasury"),
					"value": "In Cheque Treasury"
				},
				{
					"label": __("Bounced"),
					"value": "Bounced"
				},
				{
					"label": __("Collected"),
					"value": "Collected"
				},
				{
					"label": __("Under Collection"),
					"value": "Under Collection"
				},
				{
					"label": __("Endore"),
					"value": "Endore"
				},
				{
					"label": __("Paid"),
					"value": "Paid"
				},
				{
					"label": __("Transferred"),
					"value": "Transferred"
				},
				{
					"label": __("Recollected"),
					"value": "Recollected"
				},
				{
					"label": __("Refused"),
					"value": "Refused"
				},
				{
					"label": __("Funding"),
					"value": "Funding"
				}
			],
		},
		{
			fieldname: "cheque_no",
			label: __("Cheque No"),
			fieldtype: "Data",
		},
		{
			fieldname: "due_date_from",
			label: __("Due Date From"),
			fieldtype: "Date",
		},
		{
			fieldname: "due_date_to",
			label: __("Due Date To"),
			fieldtype: "Date",
		},
	],

	formatter(value, row, column, data, default_formatter) {
		// Show a button instead of the "name"
		if (column.fieldname == "button") {
			const button_html = `<button id="history-btn-${value}" class="btn btn-default btn-xs" onclick="frappe.query_reports['Cheques'].history('${value}')">History</button>`;
			value = button_html;
		}
		return default_formatter(value, row, column, data);
	},

	history(name) {
		frappe.call({
			method: 'cheque_manager.api.cheque_status',
			args: {
				'voucher': name,
			},
			callback: function (r) {
				if (!r.exc) {
					if (r.message) {
						frappe.msgprint({
							title: __('Cheque Status'),
							indicator: 'green',
							message: r.message
						});
					}
					if (r.message === 0) {
						// Remove the button if r.message is 0
						frappe.msgprint({
							title: __('Cheque Status'),
							indicator: 'red',
							message: "No History Found for this Cheque"
						});
					}
				}
			}
		});
	},


};

File: cheque_manager/cheque_manager/report/cheques/cheques.py
# Copyright (c) 2024, Mohammed Nasser and contributors
# For license information, please see license.txt

import frappe
from frappe import _
from frappe.query_builder.functions import Date
from datetime import datetime
from datetime import datetime, timedelta

Filters = frappe._dict

def execute(filters: Filters = None) -> tuple:
    columns = get_columns()
    data = get_data(filters)
    return columns, data


def get_columns() -> list[dict]:
    return [
        {"label": _("Date"), "fieldname": "posting_date", "fieldtype": "Data"},
        {"label": _("Cheque Type"), "fieldname": "cheque_type", "fieldtype": "Data"},
        {"label": _("Bank Name"), "fieldname": "bank_name", "fieldtype": "Link", "options":"Account", "width":200},
        {"label": _("Beneficiary Name"), "fieldname": "beneficiary_name", "fieldtype": "Data", "width":200},
        {"label": _("Status"), "fieldname": "status", "fieldtype": "Data", "width":150},
        {"label": _("Cheque No."), "fieldname": "cheque_no", "fieldtype": "Data", "width":150},
        {"label": _("Due Date"), "fieldname": "due_date", "fieldtype": "Data", "width":150},
        {"label": _("Voucher No."), "fieldname": "voucher_no", "fieldtype": "Link", "options":"Payment Entry", "width":200},
        {"label": _("Paid Amount"), "fieldname": "paid_amount", "fieldtype": "Currency", "width":200},
        {"label": _("History"), "fieldname": "button", "fieldtype": "Data", "width":200},
    ]

def get_data(filters: Filters) -> list[dict]:
    
    conditions = []
    if filters.get("company"):
        conditions.append(f"pe.company = '{filters.get('company')}'")
    if filters.get("from_date"):
        conditions.append(f"pe.posting_date >= '{filters.get('from_date')}'")
    if filters.get("to_date"):
        conditions.append(f"pe.posting_date <= '{filters.get('to_date')}'")
    if filters.get("cheque_type"):
        cheque_type_list = filters.get("cheque_type")
        if len(cheque_type_list) == 1:
            cheque_type = cheque_type_list[0]
            conditions.append(f"pe.mode_of_payment = '{cheque_type}'")
        else:
            cheque_type = tuple(cheque_type_list)
            conditions.append(f"pe.mode_of_payment IN {cheque_type}")
    if filters.get("bank_name"):
        bank_name_list = filters.get("bank_name")
        if len(bank_name_list) == 1:
            bank_name = bank_name_list[0]
            conditions.append(f"pe.custom_bank_name = '{bank_name}'")
        else:
            bank_name = tuple(bank_name_list)
            conditions.append(f"pe.custom_bank_name IN {bank_name}")
    if filters.get("beneficiary_name"):
        conditions.append(f"pe.party LIKE '%{filters.get('beneficiary_name')}%'")
    if filters.get("status"):
        conditions.append(f"pe.status = '{filters.get('status')}'")
    if filters.get("cheque_no"):
        conditions.append(f"pe.reference_no LIKE '%{filters.get('cheque_no')}%'")
    if filters.get("due_date_from") and filters.get("due_date_to"):
        conditions.append(f"pe.reference_date BETWEEN '{filters.get('due_date_from')}' AND '{filters.get('due_date_to')}'")
    condition_str = " AND ".join(conditions)
    if condition_str:
        condition_str = "AND " + condition_str

    sql = frappe.db.sql(f"""
        SELECT 
        pe.posting_date as posting_date,
        pe.mode_of_payment as mode_of_payment,
        pe.custom_bank_name as bank_name,
        pe.party as party,
        pe.status as status,
        pe.reference_no as reference_no,
        pe.reference_date as reference_date,
        pe.name as name,
        pe.paid_amount as paid_amount
        FROM `tabPayment Entry` pe
        WHERE pe.docstatus=1 AND pe.is_cheque=1 {condition_str}
        ORDER BY pe.posting_date ASC, pe.name ASC, pe.reference_date ASC
        """, as_dict=True)
    response = []
    for row in sql:
        response.append({
            "posting_date": row.posting_date,
            "cheque_type": row.mode_of_payment,
            "bank_name": row.bank_name,
            "beneficiary_name": row.party,
            "status": row.status,
            "cheque_no": row.reference_no,
            "due_date": row.reference_date,
            "voucher_no": row.name,
            "paid_amount": row.paid_amount,
            "button": row.name
        })
        
    return response
